src/collector/filter_collector_wrapper.rs

```diff
@@ -146,35 +173,240 @@ where
     }
 }
 
-pub struct FilterSegmentCollector<TSegmentCollector, TPredicate, TPredicateValue>
-where
-    TPredicate: 'static,
-    DynamicColumn: Into<Option<columnar::Column<TPredicateValue>>>,
-{
-    fast_field_reader: Arc<dyn ColumnValues<TPredicateValue>>,
+pub struct FilterSegmentCollector<
+    TSegmentCollector,
+    TPredicate,
+    TPredicateValue,
+    const CONJUNCTION: bool,
+> {
+    column_opt: Option<Column<TPredicateValue>>,
     segment_collector: TSegmentCollector,
     predicate: TPredicate,
     t_predicate_value: PhantomData<TPredicateValue>,
 }
 
-impl<TSegmentCollector, TPredicate, TPredicateValue> SegmentCollector
-    for FilterSegmentCollector<TSegmentCollector, TPredicate, TPredicateValue>
+impl<TSegmentCollector, TPredicate, TPredicateValue, const CONJUNCTION: bool>
+    FilterSegmentCollector<TSegmentCollector, TPredicate, TPredicateValue, CONJUNCTION>
+where
+    TPredicateValue: PartialOrd + Copy + Debug + Send + Sync + 'static,
+    TPredicate: 'static + Fn(TPredicateValue) -> bool + Send + Sync,
+{
+    #[inline]
+    fn accept_document(&self, doc_id: DocId) -> bool {
+        if let Some(column) = &self.column_opt {
+            for val in column.values_for_doc(doc_id) {
+                if CONJUNCTION ^ (self.predicate)(val) {
+                    return !CONJUNCTION;
+                }
+            }
+        }
+        CONJUNCTION
```

Comment:
I don't think this logic is correct.

Reply:
I dropped the commit, but would be glad to understand the issue.

I was thinking this would expand to

```rust
if let Some(column) = &self.column_opt {
    for val in column.values_for_doc(doc_id) {
        if false ^ (self.predicate)(val) {
            return !false;
        }
    }
}
false
```

meaning

```rust
if let Some(column) = &self.column_opt {
    for val in column.values_for_doc(doc_id) {
        if (self.predicate)(val) {
            return true;
        }
    }
}
false
```

for `CONJUNCTION == false` and similarly

```rust
if let Some(column) = &self.column_opt {
    for val in column.values_for_doc(doc_id) {
        if true ^ (self.predicate)(val) {
            return !true;
        }
    }
}
true
```

meaning

```rust
if let Some(column) = &self.column_opt {
    for val in column.values_for_doc(doc_id) {
        if !(self.predicate)(val) {
            return false;
        }
    }
}
true
```

for `CONJUNCTION == true`. Don't these loops implement "any" resp. "all" semantics?

Reply:
ah sorry yes. Somehow I read `^` as `&` my bad.